### PR TITLE
Just copy unmodified files

### PR DESCRIPTION
**Pull request by rbran**
Have the objective to speedup the lib by implementing something like https://github.com/wcampbell0x2a/backhand/issues/130

### CODE REVIEW

**Comment by rbran** [comment]
Right now it just require a function `just_copy_it` in parallel to `DataWriter::add_bytes` in case the compression of the source and destination files are the same.

This function is a little more problematic then I tough first, so I'll probably need a few days to implement it.

File: src/filesystem.rs [hunk 1]
```diff
@@ -47,175 +47,319 @@ pub struct FilesystemReader<R: SquashFsReader> {
...
+    Fragment {
+        data: &'b [u8],
+        fragment: &'a Fragment,
+    },
+    CachedFragment {
+        data: &'b [u8],
+        fragment: Fragment,
+    },
+}
+//TODO: DISCUSTING, please have mercy on my soul
```
Comment (wcampbell0x2a) [inline src/filesystem.rs:119]: lmao

**Comment by rbran** [comment]
I managed to make it work.

The command `add archlinux.sfs empty /tmp` is ~250% faster. The bench that is read/write is ~50% faster.

But the code is somewhat shabby and need some rework.

File: src/filesystem.rs (hunk 1 above)
Comment (rbran) [reply src/filesystem.rs:124]: I need to find a more elegant solution for that, this is horrible. I'll try to fix this in the next few days.

**Comment by rbran** [comment]
Maybe the real lazy evaluation was the friends we will made along the way.

Let me know if there is something that you don't like, so I can fix it.

File: src/filesystem.rs [hunk 2]
```diff
@@ -47,175 +47,305 @@ pub struct FilesystemReader<R: SquashFsReader> {
...
+        }
+    }
+}
 
+pub struct SquashfsReadFile<'a, R: ReadSeek> {
+    raw_data: SquashfsRawData<'a, R>,
+    buf_read: Vec<u8>,
+    buf_decompress: Vec<u8>,
+    //Safety: will point to one of the buffers, modify that require extra care
+    last_read: &'static [u8],
```
Comment (wcampbell0x2a) [inline src/filesystem.rs:286]: Not at my computer, but couldn't this be replaced with a `Box` for no performance loss and no unsafe?

Or `RefCell`, as I _think_ that would keep mem usage down?

Comment (rbran) [reply src/filesystem.rs]: Yes, the unsafe can be removed with little to no performance loss. No `Box` or `RefCell` required, just need to replace the reference by an index.

Comment (rbran) [reply src/filesystem.rs]: The main change being `std::mem::swap(&mut self.buf_read, &mut self.buf_decompress)` so the final data is always stored in `buf_decompress`.

Comment (rbran) [reply src/filesystem.rs]: Done on [0a36dc1](https://github.com/wcampbell0x2a/backhand/pull/132/commits/0a36dc15e6bdecc7ebf8bc1f2d3719d339926423), with some extra changes to make it possible.

**Pull request merged**